### PR TITLE
Handle comments in RFC5322 dates

### PR DESCRIPTION
**Pull request by gahr**

### CODE REVIEW

File: test/date/mutt_date_parse_date.c [hunk 1]
```diff
@@ -48,38 +48,39 @@ void test_mutt_date_parse_date(void)
...
+     { "Wed, 13 Jun 2007 12:-1:56 +0100",            -1         },
+     { "Wed, 13 Jun 2007 12:60:56 +0100",            -1         },
+     { "Wed, 13 Jun 2007 (bar baz) 12:60:56 +0100",  -1         },
+     { "Wed, 13 Jun 2007 12:34:-1 +0100",            -1         },
+     { "Wed, 13 Jun 2007 12:34:61 +0100",            -1         },
+     { "Wed, 13 Jun 2007 12:34:56 -0100",            1181741696 },
+     { "Wed, 13 Jun (Ju (06)n) 2007 12:34:56 -0100", 1181741696 },
+     { "Wed, 13 Jun 2007 12:34:56 -0100 (CET)",      1181741696 },
+     { "Wed, 13 Jun 2007 12:34:56 -0100 (CET)",      1181741696 },
+     { "Wed, 13 Jun 2007 12:34:56 (UTC)",            1181738096 },
```
Comment (flatcap) [inline test/date/mutt_date_parse_date.c:79]: This fails (but sort of doesn't).
The timezone is regex'd as `(UTC)` which can't be parsed.

> Error: Cannot parse TZ \< (UTC)\>

but that doesn't cause `mutt_date_parse_date()` to fail.

Comment (gahr) [reply test/date/mutt_date_parse_date.c:79]: That's expected - that is a comment, not a proper timezone as far as I understand the RFC. I will add a comment.

Comment (gahr) [reply test/date/mutt_date_parse_date.c]: Turns out, this is already covered by https://github.com/neomutt/neomutt/pull/2242/files/6f71d0ba5db0d8fec3c49cf07d9908db6652358c#diff-ed2732a1db2886f042ef5b242f138ddcR53

Comment (flatcap) [reply test/date/mutt_date_parse_date.c]: If it's a comment, then there's no timezone info (which is required), but parse still succeeds.

Comment (gahr) [reply test/date/mutt_date_parse_date.c]: you are right, I am a bit lax on the timezone and assume UTC. I will change that.

File: mutt/date.c [hunk 2]
```diff
@@ -468,20 +468,27 @@ time_t mutt_date_parse_date(const char *s, struct Tz *tz_out)
...
-  const regmatch_t *mtz = &match[PREX_RFC5322_DATE_MATCH_TZ];
-  const regmatch_t *mtzobs = &match[PREX_RFC5322_DATE_MATCH_TZ_OBS];
+  const regmatch_t *mday = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_DAY : PREX_RFC5322_DATE_MATCH_DAY];
+  const regmatch_t *mmonth = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_MONTH : PREX_RFC5322_DATE_MATCH_MONTH];
+  const regmatch_t *myear = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_YEAR : PREX_RFC5322_DATE_MATCH_YEAR];
+  const regmatch_t *mhour = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_HOUR : PREX_RFC5322_DATE_MATCH_HOUR];
+  const regmatch_t *mminute = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_MINUTE : PREX_RFC5322_DATE_MATCH_MINUTE];
+  const regmatch_t *msecond = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_SECOND : PREX_RFC5322_DATE_MATCH_SECOND];
+  const regmatch_t *mtz = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_TZ : PREX_RFC5322_DATE_MATCH_TZ];
+  const regmatch_t *mtzobs = &match[obsolete ?PREX_RFC5322_DATE_CFWS_MATCH_TZ_OBS : PREX_RFC5322_DATE_MATCH_TZ_OBS];
```
Comment (flatcap) [inline mutt/date.c:491]: Urgh!  I'm tempted to force the formatting:

```suggestion
  // clang-format off
  const regmatch_t *mday    = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_DAY    : PREX_RFC5322_DATE_MATCH_DAY];
  const regmatch_t *mmonth  = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_MONTH  : PREX_RFC5322_DATE_MATCH_MONTH];
  const regmatch_t *myear   = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_YEAR   : PREX_RFC5322_DATE_MATCH_YEAR];
  const regmatch_t *mhour   = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_HOUR   : PREX_RFC5322_DATE_MATCH_HOUR];
  const regmatch_t *mminute = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_MINUTE : PREX_RFC5322_DATE_MATCH_MINUTE];
  const regmatch_t *msecond = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_SECOND : PREX_RFC5322_DATE_MATCH_SECOND];
  const regmatch_t *mtz     = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_TZ     : PREX_RFC5322_DATE_MATCH_TZ];
  const regmatch_t *mtzobs  = &match[obsolete ? PREX_RFC5322_DATE_CFWS_MATCH_TZ_OBS : PREX_RFC5322_DATE_MATCH_TZ_OBS];
  // clang-format on
```

Comment (gahr) [reply mutt/date.c:491]: Definitely!

File: mutt/date.c [hunk 3]
```diff
@@ -468,20 +468,29 @@ time_t mutt_date_parse_date(const char *s, struct Tz *tz_out)
...
+  bool obsolete = false;
+
   regmatch_t *match = mutt_prex_capture(PREX_RFC5322_DATE, s);
   if (!match)
-    return -1;
+  {
+    match = mutt_prex_capture(PREX_RFC5322_DATE_CFWS, s);
+    if (!match)
+      return -1;
+    obsolete = true;
```
Comment (flatcap) [inline mutt/date.c:479]: how about a `mutt_debug()` here

File: mutt/prex.c [hunk 4]
```diff
@@ -121,29 +121,52 @@ static struct PrexStorage *prex(enum Prex which)
       PREX_RFC5322_DATE,
       PREX_RFC5322_DATE_MATCH_MAX,
       /* Spec: https://tools.ietf.org/html/rfc5322#section-3.3 */
-#define C "" //"\\([^()\\]*\\)?" // Comment are obsolete but still allowed
       "^"
         "("
-        C " *"
-          "(Mon|Tue|Wed|Thu|Fri|Sat|Sun)"  // Day of week
-        C ", "
+          "(Mon|Tue|Wed|Thu|Fri|Sat|Sun)" // Day of week
```
Comment (flatcap) [inline mutt/prex.c:126]: There are now two copies of `(Mon|Tue|...)` and three `(Jan|Feb|...)`
Should we adopt your `PREX_DOW`, `PREX_MONTH`, etc

Comment (gahr) [reply mutt/prex.c:126]: yes, that's coming in the other PR. I can bring them here or we can wait for that one to be finished.

Comment (flatcap) [reply mutt/prex.c:126]: I'll wait

**Pull request merged**